Feature refactor decoding with plugin extra

bpoplauschi opened the pull request:
@dreampiggy I have taken a deeper look at https://github.com/rs/SDWebImage/pull/1991 and I like most of your work there :)

So I decided to try some improvements on top of that. So this PR is from my fork to yours :) this was so non-trivial :))

First of all, I made a commit merging the latest master into your `feature_refactor_decoding_with_plugin` branch.

Then 2 more commits with my ideas on top of yours:
- overall idea: the singleton `SDWebImageCodersManager` holds an array of `SDWebImageCoder` (so we avoid passing them along between objects). When we want to decode/encode, we ask each coder in the array if they can decode/encode that data/format, otherwise fallback to the next. The array is reversed, so when we initially populate it with `IOCoder`, `GIFCoder` and `WebPCoder`, those will be the last ones. Every coder added by the user will have priority over our default ones.
- I wanted this change different from the `SDWebImageDecoder` approach you took because I think it's easier to add some coders to this array rather than having to replace the entire Coder object.
- `SDWebImageCoder` methods made non-optional (except `incrementalDecodedImageWithData:format:`) which can be optional - this was done so we avoid all the checks
- `SDWebImageCoder` was added 2 methods, `canDecodeData:` and `canEncodeImageFormat:`, used for customization. Basically, any coder implementing this protocol can decide which data/format can handle. (I think I should rename to `canEncodeToImageFormat:`).
- `SDWebImageCodersManager` holds the array of coders (which is returned reversed), synchronises the CRUD ops on this array and takes care of the fallback logic. For simplicity it implements the `SDWebImageCoder` protocol so everyone can pass those messages to it and under the hood, this class will take care of passing to the appropriate coder. 
- Open question: in case no coder can decode/encode for us, should we A - return nil or B fallback to the IOCoder and ignore any potential setting by the users that may have removed it from the coders array.
- no need for the `SDWebImageDecoder` class
- ran `pod install` on the `Tests` project using `CocoaPods 1.3.1`
- simplified `shouldDecode` decision from `SDWebImageDownloaderOperation` after previous code removals

dreampiggy committed:
Fix SDWebImage v4 can not import libwebp framework's header files.

Fix FLAnimatedImageView mixed local and global import

dreampiggy committed:
Performance enhancement related to single WebP image and animated WebP decoding

1. Change the code to not decode WebP images created in SD because it has been decoded
2. Add @autoreleasepool for animated WebP decoding do-while loop to reduce memory peak

bpoplauschi committed:
Attempt to fix #1883 by using a weakSelf reference to remove from callbackBlocks (on the barrierQueue)

bpoplauschi committed:
Fixed #1993 Load image with option `SDWebImageRefreshCached` get error
#1737 introduced an issue that is using NSURLRequestReturnCacheDataDontLoad for our cached images which makes us never download again, which is not ok when using `SDWebImageRefreshCached`.
Fixed by reverting to the original implementation here: `NSURLRequestUseProtocolCachePolicy` vs `NSURLRequestReloadIgnoringLocalCacheData` (when `SDWebImageRefreshCached` is set)

bpoplauschi committed:
The docs say the progressBlock is called on a background queue, so when doing UI, we need to dispatch on the main queue
* @param progressBlock  A block called while image is downloading
* @note the progress block is executed on a background queue

bpoplauschi committed:
Update for cache policy calculation: I think that if SDWebImageDownloaderIgnoreCachedResponse is there, we need NSURLRequestReloadIgnoringLocalCacheData so we always reload.

bpoplauschi committed:
Revert "Update for cache policy calculation: I think that if SDWebImageDownloaderIgnoreCachedResponse is there, we need NSURLRequestReloadIgnoringLocalCacheData so we always reload."

This reverts commit ae75b747b039ff3650e9883fd291017133218aaa.

DwayneCoussement committed:
CLANG_WARN_STRICT_PROTOTYPES: Yes please!

bpoplauschi committed:
Merge pull request #1995 from DwayneCoussement/strict-prototype

CLANG_WARN_STRICT_PROTOTYPES: Yes please!

iwill committed:
Fix issue #2001, add sd_currentBackgroundImageURL and sd_backgroundImageURLForState: for UIButton

iwill committed:
Fix issue #2001

bpoplauschi committed:
Merge pull request #2002 from iwill/master

Fix issue #2001, add sd_currentBackgroundImageURL and sd_backgroundIm…

dreampiggy committed:
Change access the imageData in NSURLSession delegate queue to solve non thread-safe issue. Also, ensure imageData in completion block is immutable to avoid accident modification from the other queue

dreampiggy committed:
Fix clang strict prototype warning in test

dreampiggy committed:
Merge pull request #1994 from bpoplauschi/fix_request_cache_policy

Fix request cache policy

dreampiggy committed:
Merge pull request #2027 from dreampiggy/fix_test_prototype_warning

Fix clang strict prototype warning in test

dreampiggy committed:
Update ManualInstallation.md (#2030)

Fix `Other Linker Flags` images

dreampiggy committed:
Add a test for SDWebImageRefreshCached

dreampiggy committed:
Merge pull request #2031 from dreampiggy/add_refresh_cache_test

Add a test for SDWebImageRefreshCached

dreampiggy committed:
Adopt the previous version behavior about SDWebImageDownloaderIgnoreCachedResponse and SDWebImageRefreshCached without hack code

dreampiggy committed:
Update macOS Demo deployment target to 10.10 to support build on Xcode 9

dreampiggy committed:
Merge pull request #2035 from dreampiggy/fix_macos_demo_deployment_target

Update macOS Demo deployment target to 10.10 to support build on Xcode 9

dreampiggy committed:
Merge pull request #2011 from dreampiggy/Fix_downloadOperation_imagedata_crash

Fix SDWebImageDownloadOperation imageData multi-thread issue.

dreampiggy committed:
Use CGImageSourceCreateIncremental to perform progressive decoding instead of decode partial data each time from the beginning to improve performance and remove that gray background

dreampiggy committed:
Merge pull request #1985 from dreampiggy/performance_webp_decoding_optimize

Performance enhancement related to single WebP and animated WebP decoding...

dreampiggy committed:
Set UIButton's placeholer-image even if the url is nil

dreampiggy committed:
Merge pull request #2043 from dreampiggy/set_uibutton_placeholder_if_url_is_nil

Set UIButton's placeholer-image even if the url is nil

dreampiggy committed:
Remove unused Image IO header

dreampiggy committed:
Merge pull request #2040 from dreampiggy/Fix_progressive_download_gray_background

Use CGImageSourceCreateIncremental to perform progressive decoding in…

bpoplauschi committed:
Merge pull request #2032 from dreampiggy/adopt_refresh_cache_behavior

Adopt the previous version behavior about SDWebImageDownloaderIgnoreC…

bpoplauschi committed:
Merge pull request #1983 from dreampiggy/fix_webp_import_issue

Fix SDWebImage v4 can not import libwebp framework's header files.

bpoplauschi committed:
Xcode 9 recommended settings applied

bpoplauschi committed:
Bumped version to 4.1.1 + updated changelog

bpoplauschi committed:
Merge branch 'master' into feature_refactor_decoding_with_plugin_extra

# Conflicts:
#	SDWebImage/FLAnimatedImage/FLAnimatedImageView+WebCache.m
#	SDWebImage/SDImageCache.m
#	SDWebImage/SDWebImageDownloaderOperation.m
#	SDWebImage/UIImage+WebP.m

bpoplauschi committed:
Simplified shouldDecode decision after previous code removals

bpoplauschi committed:
Pluginable decoders refactoring started by dreampiggy continuation (#1991):
- overall idea: the singleton SDWebImageCodersManager holds an array of SDWebImageCoder (so we avoid passing them along between objects). When we want to decode/encode, we ask each coder in the array if they can decode/encode that data/format, otherwise fallback to the next. The array is reversed, so when we initially populate it with IOCoder, GIFCoder and WebPCoder, those will be the last ones. Every coder added by the user will have priority over our default ones.
- SDWebImageCoder methods made non-optional (except incrementalDecodedImageWithData:format:) which can be optional - this was done so we avoid all the checks
- SDWebImageCoder was added 2 methods, canDecodeData: and canEncodeImageFormat:, used for customization. Basically, any coder implementing this protocol can decide which data/format can handle
- SDWebImageCodersManager holds the array of coders (which is returned reversed), synchronises the CRUD ops on this array and takes care of the fallback logic. For simplicity it implements the SDWebImageCoder protocol so everyone can pass those messages to it and under the hood, this class will take care of passing to the appropriate coder. Open question: in case no coder can decode/encode for us, should we A - return nil or B fallback to the IOCoder and ignore any potential setting by the users that may have removed it from the coders array.
- no need for the SDWebImageDecoder class
- ran `pod install` on the Tests project using CocoaPods 1.3.1

bpoplauschi commented:
PS: the changes in your branch and thus mine are backwards compatible with our existing versions (do not alter the public API), so i think we can include this refactor into 4.2.0.

dreampiggy commented on SDWebImage/SDWebImageCoder.h:
```diff
@@ -30,27 +30,30 @@ CG_EXTERN BOOL SDCGImageRefContainsAlpha(_Nullable CGImageRef imageRef);
+ @return YES if this coder can decode the data, NO otherwise
  */
-- (nullable UIImage *)decodedImageWithData:(nullable NSData *)data format:(SDImageFormat)format;
+- (BOOL)canDecodeData:(nullable NSData *)data;
```
If we want to make this more generic, we do not need to ask the image format while decoding, right ? User can use our implementation or their own `sd_imageFormatForImageData` to check which image format of the data and decide the procedure.

If we do so, these protocol method should remove that `format` args:

+ `decodedImageWithData:format`
+ `decompressedImageWithImage:data:format:shouldScaleDown`(Also I think this method last one arg a little strange. Maybe a dictionary with info key is better?)
+ `incrementalDecodedImageWithData:format`

bpoplauschi replied:
Done - see commits dc2664d and a3f45ae

bpoplauschi replied:
PS: the new signature for the decompress method is `- (UIImage *)decompressedImageWithImage:(UIImage *)image data:(NSData **)data options:(NSDictionary<NSString*, NSObject*>*)optionsDict` which is indeed generic and looks like a good API design.
We need to pass `{@"SDWebImageScaleDownLargeImages": @(YES)}` for the scaleDown to be active.

dreampiggy commented on SDWebImage/SDWebImageCodersManager.m:
```diff
@@ -0,0 +1,147 @@
+- (NSArray<SDWebImageCoder> *)coders {
+    __block NSArray<SDWebImageCoder> *sortedCoders = nil;
+    dispatch_sync(self.mutableCodersAccessQueue, ^{
+        sortedCoders = (NSArray<SDWebImageCoder> *)[[[self.mutableCoders copy] reverseObjectEnumerator] allObjects];
```
You're right. To avoid that user remove all the coders built-in. We need a nil check here and return the default coders array.

bpoplauschi replied:
I don't agree :) if the user wants to remove all the coders, so be it. Not our problem :)

dreampiggy commented:
Thanks for your work!
After consideration, previously the solution is to let user to create their own id<SDWebImageCoder> to dispatch the actual method. But it seems not so `pluggable` since you need to create class and set to the coders for each `Downloader` `Manager` `Cache` shared instance.

Your idea about this plugin is more elegant and it looks the same like some Web backend middleware. (Ask each middleware to decide whether it can process). I think this one is better.

dreampiggy commented:
Moreover, for that `SDWebImageCoder` protocol, I advice some more changes.

1. Since we add one `canDecodeData`, user can use their own method to detect image format or use our `sd_imageFormatForImageData`. So we do not need that `SDImageFormat` arg in each method declaration.
2. For `encodedDataWithImage:format`, can we add one more arg with a dictionary such as `properties`? Because UIImage just keep the image bitmap but not image EXIF metadata. EXIF metadata keep the more informatation such as orientation, capture data, location...etc.
In https://github.com/rs/SDWebImage/pull/2034 I add two more method to fix that `UIImagePNGRepresentation` to preserve image orientation and EXIF information. I think this can let the encoding more generic.
And for this, maybe we should also add a extra method to let user grab the properties from the image data. Because not all the image format is supported in Image/IO such as WebP...

bpoplauschi committed:
Attempt to fix the tests (test40InsertionOfImageData still fails locally even if I fixed it: replaced will.beTruthy to equal and make sure we wait for the 2 async calls to end). Also properly used the SDWebImageTestDecoder

bpoplauschi committed:
Removed `format` from the decode methods of `SDWebImageCoder`
Improvement: instead of `incrementalDecodedImageWithData` being an optional method of `SDWebImageCoder`, decided to create `SDWebImageProgressiveCoder` so it's easier to check for this protocol instead of respondsToSelector

bpoplauschi committed:
API improvement for `decompressedImageWithImage`: the `shouldScaleDown` param was replaced by a generic dict `options` which can contain any options, that way it is more generic. For shouldScaleDown options pass: {@"SDWebImageScaleDownLargeImages": @(YES)}

bpoplauschi committed:
Per https://github.com/rs/SDWebImage/pull/2034, trying to add `propertiesOfImageData` method to our coder protocol + pass the properties to the `encodedDataWithImage` call. I'm not sure about this, looks weird since we only call if from a place where the data is nil.

bpoplauschi committed:
Final fix for our tests: get the NSData through UIImage

bpoplauschi commented:
> Moreover, for that `SDWebImageCoder` protocol, I advice some more changes.
> 1. Since we add one `canDecodeData`, user can use their own method to detect image format or use our `sd_imageFormatForImageData`. So we do not need that `SDImageFormat` arg in each method declaration.

Done - see dc2664d

> 2. For `encodedDataWithImage:format`, can we add one more arg with a dictionary such as `properties`? Because UIImage just keep the image bitmap but not image EXIF metadata. EXIF metadata keep the more informatation such as orientation, capture data, location...etc.
In https://github.com/rs/SDWebImage/pull/2034 I add two more method to fix that `UIImagePNGRepresentation` to preserve image orientation and EXIF information. I think this can let the encoding more generic.
> And for this, maybe we should also add a extra method to let user grab the properties from the image data. Because not all the image format is supported in Image/IO such as WebP...

I've attempted this proposal in this commit 786fc7f.
So I added `propertiesOfImageData` method to our coder protocol + pass the properties to the `encodedDataWithImage` call. 
I'm not sure about this, looks weird since we only call if from a place where the data is nil thus we can get 0 properties :)
See SDImageCache.m L220:
```
if (!data && image) {
    SDImageFormat imageFormatFromData = [NSData sd_imageFormatForImageData:data];
    NSDictionary *properties = [[SDWebImageCodersManager sharedInstance] propertiesOfImageData:data];
    data = [[SDWebImageCodersManager sharedInstance] encodedDataWithImage:image format:imageFormatFromData properties:properties];
}
```
Maybe you can take it from here.

bpoplauschi commented:
@dreampiggy take a look when you can and let me know

dreampiggy commented:
Thanks! I'll merge all the commits into this repo's branch and I think you can directly push to this branch if you find something to modify.

dreampiggy commented:
@bpoplauschi I think my consideration for that image `properties` is too away from most users need. Many user do not need the total EXIF metadata but only want to keep the image orientation. Since total EXIF data it's used for advanced photography software but not a common case. So I wonder why we need to preserve all the metadata and increase the complexity 😕 

A better way is to just grap the image orientation from `UIImage.imageOrientation` and save this property. So we do not need that `properties` arg and that `propertiesOfImageData`(Sorry, I removed it :D). And this can keep the coder protocol more easier.
